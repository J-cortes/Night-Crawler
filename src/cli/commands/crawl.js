// @flow

import { requireCrawler, writeJSON } from '../util';
import formatterFactory from '../formatters';
import ProgressBar from 'ascii-progress';
import { FailedAnalysisError } from '../errors';
import type yargs from 'yargs';

exports.command = 'crawl [crawlerfile]';
exports.describe = 'execute the crawl defined in the active config file';
exports.builder = (yargs: yargs) => {
  yargs.positional('crawlerfile', {
    describe: 'the name of the crawler file',
    default: 'nightcrawler.js'
  });
  yargs.option('silent', {
    alias: 'n',
    describe: 'silence all output',
    type: 'boolean',
    default: false
  });
  yargs.option('output', {
    alias: 'o',
    describe: 'filename to write JSON report to'
  });
  yargs.option('format', {
    alias: 'f',
    describe: 'the format to output analysis result in',
    default: 'console'
  });
  yargs.option('formatOptions', {
    alias: 'g',
    describe: 'formatter options, in JSON format'
  });
};
exports.handler = async function(argv: Object) {
  const { crawlerfile, silent, output, format } = argv;
  const formatOptions = argv.formatOptions
    ? JSON.parse(argv.formatOptions)
    : {};
  const crawler = requireCrawler(crawlerfile);

  let progress;
  let start = new Date();

  if (!silent) {
    console.log(`Starting crawl`);
    progress = new ProgressBar({ total: 1 });
    crawler.on('response', () => {
      progress.total = crawler.queue.length;
      progress.tick();
    });
  }

  const data = await crawler.crawl();

  if (!silent) {
    let elapsed = Math.round((new Date() - start) / 1000);
    console.log(`Crawled ${crawler.queue.length} urls in ${elapsed} seconds.`);
  }

  if (output) {
    writeJSON(output, data);
  }

  const analysis = await crawler.analyze(data);

  if (format) {
    const formatter = formatterFactory(format, formatOptions);
    let formattedOutput = formatter.format(analysis);
    if (formattedOutput) {
      console.log(formattedOutput);
    }
  }

  if (analysis.hasFailures()) {
    throw new FailedAnalysisError('Analysis reported an error');
  }
};
