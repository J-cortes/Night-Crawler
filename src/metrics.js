
function averageTime(collection) {
    return collection.reduce((sum, i) => sum + i.data.time, 0) / collection.length;
}
averageTime.displayName = 'Average Backend Time';

function percentSuccess(collection) {
    return collection.filter(i => i.data.status <= 500).length / collection.length * 100;
}
percentSuccess.displayName = '% Successful';

function totalRequests(collection) {
    return collection.length;
}
totalRequests.displayName = 'Total Requests';

module.exports = {
    averageTime,
    percentSuccess,
    totalRequests
}