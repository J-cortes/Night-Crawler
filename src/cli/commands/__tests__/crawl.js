import { handler } from '../crawl';
import stream from 'stream';
import os from 'os'
import fs from 'fs'
import Crawler from '../../../crawler';
import DummyDriver from '../../../driver/dummy';
import { Number } from '../../../metrics';
import { FailedAnalysisError } from '../../errors';
import JUnitFormatter from '../../formatters/junit';

jest.mock('../../formatters/junit')

describe('Crawl Command', function() {
  it('Executes the crawl', function() {
    let called = 0;
    const crawler = new Crawler();
    crawler.on('setup', () => called++);

    return handler({
      crawlerfile: crawler,
      stdout: new stream.PassThrough()
    }).then(function() {
      expect(called).toEqual(1);
    });
  });

  it('Displays output to indicate the success of the crawl', function() {
    var stdout = new stream.PassThrough();

    const crawler = new Crawler('', new DummyDriver());
    crawler.on('setup', () => {
      crawler.enqueue('http://example.com/');
      crawler.enqueue('http://example.com/');
    });

    return handler({
      crawlerfile: crawler,
      stdout
    }).then(function() {
      expect(stdout.read().toString()).toMatchSnapshot();
    });
  });

  it('Outputs analysis at the end of the crawl if requested', function() {
    var stdout = new stream.PassThrough();

    const crawler = new Crawler('', new DummyDriver());
    crawler.on('analyze', function(r, a) {
      a.addMetric('foo', new Number('MyTestMetric', 0, 1));
    });

    return handler({
      crawlerfile: crawler,
      stdout
    }).then(function() {
      expect(stdout.read().toString()).toContain('MyTestMetric');
    });
  });

  it('Throws an error if the analysis contains failures', function() {
    const crawler = new Crawler('', new DummyDriver());
    crawler.on('analyze', function(r, a) {
      a.addMetric('foo', new Number('MyTestMetric', 2, 1));
    });

    let called = 0;
    return handler({
      crawlerfile: crawler,
      stdout: new stream.PassThrough()
    })
      .catch(function(err) {
        called++;
        expect(err).toBeInstanceOf(FailedAnalysisError);
      })
      .then(function() {
        expect(called).toEqual(1);
      });
  });

  it('Should save a junit report if requested', function() {
      var filename = `${os.tmpdir()}/nightcrawler-${Math.floor((Math.random() * 10000))}`;
      const crawler = new Crawler('', new DummyDriver);

      return handler({
          crawlerfile: crawler,
          junit: filename
      })
      .then(function() {
        expect(JUnitFormatter).toHaveBeenCalledTimes(1)
        expect(JUnitFormatter.mock.calls[0]).toEqual([filename])
      })
  });

  it('Should save a valid JSON run report if requested', function() {
    var filename = `${os.tmpdir()}/nightcrawler-${Math.floor((Math.random() * 10000))}`;
    const crawler = new Crawler('', new DummyDriver);
    return handler({
        crawlerfile: crawler,
        json: filename
    }).then(function() {
      expect(fs.existsSync(filename)).toEqual(true)
      fs.unlinkSync(filename)
    })
  });
});
