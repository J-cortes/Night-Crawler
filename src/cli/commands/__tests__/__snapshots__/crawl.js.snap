// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crawl Command Displays output to indicate the success of the crawl 1`] = `
"Setup
Crawl
Analyze
Results
======
[33mNo Results[39m

Metrics
=======
[33mNo Results[39m
"
`;
